pandas/indexes/base.py

```diff
@@ -2366,7 +2365,17 @@ def groupby(self, to_groupby):
         groups : dict
             {group name -> group labels}
         """
-        return self._groupby(self.values, _values_from_object(to_groupby))
+        from pandas.core.groupby import _groupby_indices
+
+        # TODO: if we are a MultiIndex, we can do better
+        # that converting to tuples
+        to_groupby = getattr(to_groupby, 'values', to_groupby)
+        result = _groupby_indices(to_groupby)
+
+        # map to the label
+        result = {k: self.take(v) for k, v in compat.iteritems(result)}
```

Comment:
AFAICT, this is *only* used on ``grouped.groups``, so by-definition we need to box. (and that's the ONLY benchmark that changed with this PR). I though this would have broader implications (on the good side), but no-go.